libs/remote_storage/src/s3_bucket.rs

```diff
@@ -279,13 +281,36 @@ impl<S: AsyncRead> AsyncRead for RatelimitedAsyncRead<S> {
     }
 }
 
+impl S3Bucket {
+    async fn wait_permit(
+        &self,
+        cancel: &CancellationToken,
+    ) -> anyhow::Result<tokio::sync::SemaphorePermit> {
+        tokio::select! {
```

Comment:
These should be biased, and always check the cancellation first. Otherwise we will have detaches behaving "randomly".